lib/Ceilometer/Client.hs

```diff
@@ -0,0 +1,129 @@
+--
+-- The code in this file, and the program it is a part of, is
+-- made available to you by its authors as open source software:
+-- you can redistribute it and/or modify it under the terms of
+-- the 3-clause BSD licence.
+--
+--
+-- This module defines a convienient interface for clients
+-- of Ceilometer.
+--
+-- For flexibility use the Collector and Fold modules.
+--
+module Ceilometer.Client
+  ( -- * Combo
+    decodeAndFold
+    -- * Decoding
+  , decode, decodeWith
+    -- * Folding
+  , foldDecoded, foldDecodedWith
+    -- * Re-exports
+  , module C
+  ) where
+
+import           Control.Applicative
+import           Control.Lens
+import           Control.Monad
+import           Data.Maybe
+import qualified Data.Traversable    as T
+import           Data.Typeable
+import           Data.Word
+import           Pipes
+
+import           Ceilometer.Fold     as C
+import           Ceilometer.Infer    as C
+import           Ceilometer.Types    as C
+import           Vaultaire.Types
+
+
+-- | Decode and fold a stream of @SimplePoint@.
+--
+--   If the assertion that these points should be type `a` is wrong: run-time error.
+--
+--   If the assertion is true but the @SourceDict@ doesn't match what we expect
+--   for the fold: `Nothing`.
+--
+--   Otherwise we get a `FoldResult` matching that type.
+--
+--   Example usage: `decodeAndFold (undefined :: proxy PDCPU) env points`
+--
+decodeAndFold
+  :: forall a m proxy. (Typeable a, Monad m, Applicative m)
+  => proxy a                    -- ^ We expect these @SimplePoint@ to be of type `a`
+  -> Env                        -- ^ @SourceDict@ to verify the above claim.
+  -> Producer SimplePoint m ()  -- ^ The raw data points to parse and aggregate.
+  -> m (Maybe (FoldResult a))   -- ^ Result
+decodeAndFold _ env points
+  = let decoded :: Maybe (Producer (Timed a) m ())
+        decoded =  useDecoding <$> decode env
+    in  fromMaybe (return Nothing) $ foldDecoded env <$> decoded
+    where useDecoding d = points >-> d >-> handleParseFails
+
+
+-- Decode ----------------------------------------------------------------------
+
+-- | Figures out what prism to use, then decode.
+--
+decode
+  :: (Typeable a, Monad m)
+  => Env                                             -- ^ @SourceDict@ to use and guess the correct decoding
+  -> Maybe (Pipe SimplePoint (Maybe (Timed a)) m ()) -- ^ Decoding pipe
+decode env = case inferPrism env of
+  Just p  -> Just $ decodeWith $ clonePrism p
+  Nothing -> Nothing -- note: don't try to fmap. ImpredicativeTypes nastiness.
+
+-- | Decodes the raw stream of SimplePoint using the given prism for the payload.
+decodeWith
+  :: Monad m
+  => Prism' Word64 a                         -- ^ Use this particular prism for decoding
+  -> Pipe SimplePoint (Maybe (Timed a)) m () -- ^ Decoding pipe
+decodeWith p = forever $ do
+  SimplePoint _ (TimeStamp t) v <- await
+  yield $ T.sequence $ Timed t $ v ^? p
+
+ignore :: Monad m => Pipe (Maybe x) x m r
```

Comment:
I can't find any difference between your definition and concat other than it is more specific.

It "behaves" the same, too.

```haskell
λ P.toList $ each [Just 1, Nothing, Just 2] >-> ignore
[1,2]
λ P.toList $ each [Just 1, Nothing, Just 2] >-> P.concat
[1,2]
```